File: app/src/main/java/com/agrogames/islandsofwar/engine/impl/unit/SmallShip.java
package com.agrogames.islandsofwar.engine.impl.unit;

import android.os.Build;

import androidx.annotation.RequiresApi;

import com.agrogames.islandsofwar.common.M;
import com.agrogames.islandsofwar.engine.abs.bullet.BulletAdder;
import com.agrogames.islandsofwar.engine.abs.common.Cell;
import com.agrogames.islandsofwar.engine.abs.common.Point;
import com.agrogames.islandsofwar.engine.abs.map.MapObject;
import com.agrogames.islandsofwar.engine.abs.map.MapProvider;
import com.agrogames.islandsofwar.engine.abs.movable.MovableObject;
import com.agrogames.islandsofwar.engine.abs.transport.TransportUnit;
import com.agrogames.islandsofwar.engine.abs.unit.UnitAdder;
import com.agrogames.islandsofwar.factories.UnitFactory;
import com.agrogames.islandsofwar.types.UnitType;
import com.agrogames.islandsofwar.engine.abs.weapon.Weapon;
import com.agrogames.islandsofwar.engine.impl.navigator.Navigator;
import com.agrogames.islandsofwar.map.abs.MapParams;
import com.agrogames.islandsofwar.map.impl.Water;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Stack;

public class SmallShip extends Unit{
    private final Stack<Cell> route = new Stack<>();
    private Cell goal;
    private boolean isMoving;
    public TransportUnit unit;

    public SmallShip(UnitType type, Cell location, Weapon[] weapons, int health, float speed, float rotationSpeed) {
        super(type, location, weapons, health, speed, rotationSpeed);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Override
    public void update(MapProvider provider, BulletAdder bulletAdder, UnitAdder unitAdder, float deltaTime) {
        if(goal != null){
            buildRoute(provider.getAll(), goal);
            goal = null;
        }
        rotate(deltaTime);
        move(provider.getAll(), unitAdder, deltaTime);
        for (Weapon weapon: getWeapons()){
            weapon.update(provider, bulletAdder, unitAdder, deltaTime);
        }
    }

    private void buildRoute(MapObject[] all, Cell goal){
        List<Cell> map = new ArrayList<>();
        for (MapObject object: all){
            if(!(object instanceof SmallShip)){
                for (Cell c: object.getTerritory()){
                    map.remove(c);
                }
            }
        }
        map.remove(goal);
        Cell[] r = new Cell[0];
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            r = Navigator.buildRoute(location, goal, map.toArray(new Cell[0]));
        }
        route.clear();
        for (int i = r.length - 1; i >= 0; i--) {
            Cell c = r[i];
            route.push(c);
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private void move(MapObject[] all, UnitAdder adder, float deltaTime){
        if(route.isEmpty()){
            isMoving = false;
            return;
        }
        if(new Cell(location).equals(route.lastElement())){
            route.pop();
        }
        if(route.isEmpty()){
            isMoving = false;
            return;
        }

        Cell g = route.lastElement();
        setRotation(g);
        Point next = next(deltaTime, g);
        Cell nextCell = new Cell(next);
        Point further = next(0.5f / speed, g);
        Cell furtherCell = new Cell(further);
        if(isTaken(nextCell, all)) {
            stop(nextCell, all, adder);
            return;
        }
        if(isTaken(furtherCell, all)) {
            stop(furtherCell, all, adder);
            return;
        }
        location = next;
    }

    private void stop(Cell taken, MapObject[] all, UnitAdder adder){
        boolean isTakingMoving = false;
        boolean stoppedOnUnit = false;
        for (MapObject object: all){
            for (Cell cell: object.getTerritory()){
                if(cell.equals(taken)){
                    if(object != this && object instanceof SmallShip){
                        isTakingMoving = object.isMoving();
                    }
                    stoppedOnUnit = true;
                }
            }
        }
        if(adder != null && !stoppedOnUnit){
            land(adder, taken);
        }
        if (isTakingMoving){
            isMoving = false;
        } else {
            if(!route.isEmpty()) buildRoute(all, route.firstElement());
        }
    }

    private void land(UnitAdder adder, Cell cell){
        if(unit == null) return;
        com.agrogames.islandsofwar.engine.abs.unit.Unit u = null;
        switch (unit.type){
            case Tank:
                u = UnitFactory.Tank(cell.x, cell.y);
                break;
            case RocketLauncher:
                u = UnitFactory.RocketLauncher(cell.x, cell.y);
                break;
        }
        if(u == null) return;
        if(!route.isEmpty() && !cell.equals(route.firstElement())){
            ((MovableObject)u).setGoal(route.firstElement());
        }
        adder.addUnit(u);
        unit = null;
        health.current = 0;
        timeSinceDestroyed = 100;
    }

    private boolean isTaken(Cell cell, MapObject[] all){
        List<Cell> map = new ArrayList<>();
        for (int x = 0; x < MapParams.Width; x++){
            for (int y = 0; y < MapParams.Height; y++){
                map.add(new Cell(x, y));
            }
        }
        for (MapObject object: all){
            if(object instanceof Water){
                for (Cell c: object.getTerritory()){
                    map.remove(c);
                }
            } else if (object != this){
                Collections.addAll(map, object.getTerritory());
            }
        }

        return map.contains(cell);
    }

    private Point next(float deltaTime, Cell goal){
        Point g = new Point(goal);
        double lx = g.x - location.x;
        double ly = g.y - location.y;
        double l = Math.sqrt(lx * lx + ly * ly);
        double rel = l / (speed * deltaTime);
        double dx = lx / rel;
        double dy = ly / rel;
        if(M.module(dx) > M.module(lx)){
            dx = lx;
            dy = ly;
        }
        if(Double.isNaN(dx)){
            dx = 0;
        }
        if(Double.isNaN(dy)){
            dy = 0;
        }
        return new Point(location.x + (float)dx,  location.y + (float) dy);
    }

    @Override
    public boolean isMoving() {
        return isMoving;
    }

    @Override
    public int getHeight() {
        return 1;
    }

    @Override
    public void setGoal(Cell goal) {
        this.goal = goal;
    }
}

File: old/app/src/main/java/ru/agrogames/islands/engine/impl/unit/SmallShip.kt
package ru.agrogames.islands.engine.impl.unit

import ru.agrogames.islands.common.M
import ru.agrogames.islands.engine.abs.bullet.BulletAdder
import ru.agrogames.islands.engine.abs.common.Cell
import ru.agrogames.islands.engine.abs.common.Point
import ru.agrogames.islands.engine.abs.graphics.GraphicsAdder
import ru.agrogames.islands.engine.abs.map.MapObject
import ru.agrogames.islands.engine.abs.map.MapProvider
import ru.agrogames.islands.engine.abs.movable.MovableObject
import ru.agrogames.islands.engine.abs.transport.TransportUnit
import ru.agrogames.islands.engine.abs.unit.IUnitAdder
import ru.agrogames.islands.engine.abs.weapon.IWeapon
import ru.agrogames.islands.engine.impl.navigator.Navigator
import ru.agrogames.islands.map.MapParams
import ru.agrogames.islands.map.Water
import java.util.Collections
import java.util.Stack

class SmallShip(texture: String, location: Cell, weapons: Array<IWeapon>, health: Int, speed: Float, rotationSpeed: Float)
    : Unit(texture, location, weapons, health, speed, rotationSpeed) {
    override val route = Stack<Cell>()
    var goal: Cell? = null
    override var isMoving = false
        private set
    var unit: TransportUnit? = null

    override fun update(provider: MapProvider, bulletAdder: BulletAdder, unitAdder: IUnitAdder, graphicsAdder: GraphicsAdder, deltaTime: Float) {
        if (goal != null) {
            buildRoute(provider.all, goal)
            goal = null
        }
        rotate(deltaTime)
        move(provider.all, unitAdder, deltaTime)
        for (weapon in weapons) {
            weapon.update(provider, bulletAdder, unitAdder, graphicsAdder, deltaTime)
        }
    }

    override fun cleverSetGoal(goal: Cell) {
        this.goal = goal
    }

    private fun buildRoute(all: Array<MapObject>, goal: Cell?) {
        val map: MutableList<Cell> = ArrayList()
        for (mo in all) {
            if ( mo !is SmallShip) {
                for (c in mo.territory) {
                    map.remove(c)
                }
            }
        }
        map.remove(goal)
        val r = Navigator.buildRoute(location, goal!!, map.toTypedArray())
        route.clear()
        for (i in r.indices.reversed()) {
            val c = r[i]
            route.push(c)
        }
    }

    private fun move(all: Array<MapObject>, adder: IUnitAdder, deltaTime: Float) {
        if (route.isEmpty()) {
            isMoving = false
            return
        }
        if (Cell(location) == route.lastElement()) {
            route.pop()
        }
        if (route.isEmpty()) {
            isMoving = false
            return
        }
        val g = route.lastElement()
        setRotation(g!!)
        val next = next(deltaTime, g)
        val nextCell = Cell(next)
        val further = next(0.5f / speed, g)
        val furtherCell = Cell(further)
        if (isTaken(nextCell, all)) {
            stop(nextCell, all, adder)
            return
        }
        if (isTaken(furtherCell, all)) {
            stop(furtherCell, all, adder)
            return
        }
        location = next
    }

    private fun stop(taken: Cell, all: Array<MapObject>, adder: IUnitAdder) {
        var isTakingMoving = false
        var stoppedOnUnit = false
        for (`object` in all) {
            for (cell in `object`.territory) {
                if (cell == taken) {
                    if (`object` !== this && `object` is SmallShip) {
                        isTakingMoving = `object`.isMoving
                    }
                    stoppedOnUnit = true
                }
            }
        }
        if (!stoppedOnUnit) {
            land(adder, taken)
        }
        if (isTakingMoving) {
            isMoving = false
        } else {
            if (!route.isEmpty()) buildRoute(all, route.firstElement())
        }
    }

    private fun land(adder: IUnitAdder, cell: Cell) {
        if (unit == null) return
        val u = unit!!.create.apply(cell)
        if (!route.isEmpty() && cell != route.firstElement()) {
            (u as MovableObject).cleverSetGoal(route.firstElement())
        }
        adder.addUnit(u)
        unit = null
        health.current = 0
        timeSinceDestroyed = 100f
    }

    private fun isTaken(cell: Cell, all: Array<MapObject>): Boolean {
        val map: MutableList<Cell?> = ArrayList()
        for (x in 0 until MapParams.width) {
            for (y in 0 until MapParams.height) {
                map.add(Cell(x, y))
            }
        }
        for (`object` in all) {
            if (`object` is Water) {
                for (c in `object`.territory) {
                    map.remove(c)
                }
            } else if (`object` !== this) {
                Collections.addAll(map, *`object`.territory)
            }
        }
        return map.contains(cell)
    }

    private fun next(deltaTime: Float, goal: Cell?): Point {
        val g = Point(goal)
        val lx = (g.x - location.x).toDouble()
        val ly = (g.y - location.y).toDouble()
        val l = Math.sqrt(lx * lx + ly * ly)
        val rel = l / (speed * deltaTime)
        var dx = lx / rel
        var dy = ly / rel
        if (M.module(dx) > M.module(lx)) {
            dx = lx
            dy = ly
        }
        if (java.lang.Double.isNaN(dx)) {
            dx = 0.0
        }
        if (java.lang.Double.isNaN(dy)) {
            dy = 0.0
        }
        return Point(location.x + dx.toFloat(), location.y + dy.toFloat())
    }

    override val height: Int
        get() = 1
}